### DISCUSSION TITLE
Asset vs. Op Graphs

### DISCUSSION THREAD

**Discussion by JSteinbauer** [dagster-io/dagster#10512] [Q&A]
Hi there,
here is a question that caused confusion in our whole data team (and I think that a corresponding clarification in you documentation would be a huge help for anyone eager to learn dagster):

With asset graphs and "normal" graphs, there are two different ways to define graphs in dagster. What makes things even more confusing is that apparently, it's possible to create graphs including assets, which by themselfes are interconnected as graphs through their arguments. Using SourceAssets, it's even possible to construct these graphs such that they don't match.

My question: I'd like to use assets as "cached building blocks with data validation", and recombine them in various data-pipelines (=graphs). Is that possible? If not, is there any "dagstonic" way to approach this problem?

Thanks a lot for your awesome support!
Best,
Jakob

Ps: This ticket is related to https://github.com/dagster-io/dagster/discussions/9687

**Comment by sryza** [comment]
Hey @JSteinbauer - when you say "recombine them", are you imagining that you'd have different topologies in different data pipelines? Like the same asset might have different dependencies depending on the pipeline? Or just that different pipelines would execute different subsets of your asset graph?

**Reply by JSteinbauer** [reply]
@Sryza yes exactly like _"Like the same asset might have different dependencies depending on the pipeline"_.
I reckonned that, with the same topology, this issue could be solved by using different assets with different function names which share the same <NAME> in @asset(name=<NAME>). But we will change the topology - for instance by including/excluding modifiers (which we'd also like to implement as assets, to store their resulting data).

**Reply by sryza** [reply]
I think I understand 75% of what you're saying, but haven't grokked the entire thing yet. Do you have an idea of what the ideal Python API would look like for expressing what you're trying to express?

**Reply by JSteinbauer** [reply]
Thanks a lot for sharing the article below - it's really nice (you might consider adding parts of it to your documentation). So as far as I understand from your documentation and the article, assets are just not supposed to be used with changing upstream dependencies and we should thus use ops instead? 

In this case, I would like to know if it's possible to cache the output of the Ops: I've seen the example where you "manually persist" the output of an op
`@op`
`def my_materialization_op(context):`
`    df = read_df()`
`    remote_storage_path = persist_to_storage(df)`
`    context.log_event(`
`        AssetMaterialization(`
`            asset_key="my_dataset", description="Persisted result to storage"`
`        )`
`    )`
`    return remote_storage_path`

but doing this I'd seem to loose any advantage of automatic data-persistence. Is there a better way to do this in dagster?

Best, 
Jakob

**Reply by sryza** [reply]
Glad you enjoyed the blog post!

If two different op outputs had used different upstream dependencies, are there situations where you'd want to cache them in the same location? The way I've been looking at it, if you want to cache something, you'd also want it to have a fixed set of dependencies.

**Comment by sryza** [comment]
This isn't a complete answer to your question, but I wonder if it's helpful? https://dagster.io/blog/roman-roads-assets-ops
